spec/models/travel_request_spec.rb

```diff
@@ -15,5 +16,131 @@
+      expect { travel_request.approve!(agent: user) }.to raise_error AASM::InvalidTransition
+    end
+
+    it "Does not approve a denied event" do
+      travel_request.deny!(agent: user)
+      expect { travel_request.approve!(agent: user) }.to raise_error AASM::InvalidTransition
+    end
+  end
+
+  describe "#deny!" do
+    let(:travel_request) { FactoryBot.create :travel_request }
+    it "denies a pending event" do
+      expect do
+        travel_request.deny!(agent: user)
+      end.to change(StateChange, :count).by(1)
+      expect(travel_request.status).to eq("denied")
+      expect(travel_request.persisted?).to be_truthy
+    end
+
+    it "Does not deny a change_request event" do
+      travel_request.change_request!(agent: user)
+      expect { travel_request.deny!(agent: user) }.to raise_error AASM::InvalidTransition
+    end
+
+    it "Does not deny a cancled event" do
+      travel_request.cancel!(agent: user)
+      expect { travel_request.deny!(agent: user) }.to raise_error AASM::InvalidTransition
+    end
+
+    it "Does not deny an approved event" do
+      travel_request.approve!(agent: user.department.head)
+      expect { travel_request.deny!(agent: user) }.to raise_error AASM::InvalidTransition
+    end
+  end
+
+  describe "#change_request!" do
+    let(:travel_request) { FactoryBot.create :travel_request }
+    it "records a change_request event" do
+      expect do
+        travel_request.change_request!(agent: user)
+      end.to change(StateChange, :count).by(1)
+      expect(travel_request.status).to eq("changes_requested")
+      expect(travel_request.persisted?).to be_truthy
+    end
+
+    it "Does not change_request an approved event" do
+      travel_request.approve!(agent: user.department.head)
+      expect { travel_request.change_request!(agent: user) }.to raise_error AASM::InvalidTransition
+    end
+
+    it "Does not change_request a canceled event" do
+      travel_request.cancel!(agent: user)
+      expect { travel_request.change_request!(agent: user) }.to raise_error AASM::InvalidTransition
+    end
+
+    it "Does not change_request a denied event" do
+      travel_request.deny!(agent: user)
+      expect { travel_request.change_request!(agent: user) }.to raise_error AASM::InvalidTransition
+    end
+  end
+
+  describe "#cancel!" do
+    let(:travel_request) { FactoryBot.create :travel_request }
+    it "cancels a pending event" do
+      expect do
+        travel_request.cancel!(agent: user)
+      end.to change(StateChange, :count).by(1)
+      expect(travel_request.status).to eq("canceled")
+      expect(travel_request.persisted?).to be_truthy
+    end
+
+    it "cancels an approved event" do
+      travel_request.approve!(agent: user.department.head)
+      travel_request.cancel!(agent: user)
+      expect(travel_request.status).to eq("canceled")
+      expect(travel_request.persisted?).to be_truthy
+    end
+
+    it "Does not cancel a change_request event" do
```

Comment:
I wonder if we should allow users to cancel change_request events if they cannot accommodate the requested changes?

Reply:
That is a good question.  We did not have it in the list, but it seems like a good addition.